File: overlays/pkg/local/gpad2mouse/Sources/gpad2mouse/StatusBar.swift
import AppKit
import GameController

class StatusBar {
    private var statusItem: NSStatusItem!
    private var enabledItem: NSMenuItem!
    private var controllerItem: NSMenuItem!
    private var statusTimer: DispatchSourceTimer?
    private let settingsWindowController = SettingsWindowController()

    private(set) var isEnabled = true

    weak var settings: Settings?
    weak var gamepadManager: GamepadManager?

    func setup() {
        statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)

        if let button = statusItem.button {
            button.image = NSImage(
                systemSymbolName: "gamecontroller.fill",
                accessibilityDescription: "gpad2mouse"
            )
        }

        let menu = NSMenu()

        controllerItem = NSMenuItem(title: "No controller", action: nil, keyEquivalent: "")
        controllerItem.isEnabled = false
        menu.addItem(controllerItem)

        menu.addItem(NSMenuItem.separator())

        enabledItem = NSMenuItem(title: "Enabled", action: #selector(toggleEnabled), keyEquivalent: "")
        enabledItem.target = self
        enabledItem.state = .on
        menu.addItem(enabledItem)

        let settingsItem = NSMenuItem(title: "Settings...", action: #selector(openSettings), keyEquivalent: ",")
        settingsItem.target = self
        menu.addItem(settingsItem)

        menu.addItem(NSMenuItem.separator())

        let quitItem = NSMenuItem(title: "Quit", action: #selector(quit), keyEquivalent: "q")
        quitItem.target = self
        menu.addItem(quitItem)

        statusItem.menu = menu

        statusTimer = DispatchSource.makeTimerSource(queue: .main) as? DispatchSourceTimer
        statusTimer?.schedule(deadline: .now(), repeating: 2.0)
        statusTimer?.setEventHandler { [weak self] in
            self?.updateControllerStatus()
        }
        statusTimer?.resume()
    }

    private func updateControllerStatus() {
        if let gc = gamepadManager?.controller {
            controllerItem.title = gc.vendorName ?? gc.productCategory
        } else {
            controllerItem.title = "No controller"
        }
    }

    @objc private func toggleEnabled() {
        isEnabled = !isEnabled
        enabledItem.state = isEnabled ? .on : .off

        if let button = statusItem.button {
            button.image = NSImage(
                systemSymbolName: isEnabled ? "gamecontroller.fill" : "gamecontroller",
                accessibilityDescription: "gpad2mouse"
            )
        }
        fputs("gpad2mouse: \(isEnabled ? "enabled" : "disabled")\n", stderr)
    }

    @objc private func openSettings() {
        guard let settings = settings, let gamepadManager = gamepadManager else { return }
        settingsWindowController.show(settings: settings, gamepadManager: gamepadManager)
    }

    @objc private func quit() {
        GCController.stopWirelessControllerDiscovery()
        fputs("gpad2mouse: shutting down\n", stderr)
        NSApp.terminate(nil)
    }
}

File: overlays/pkg/local/gpad2mouse/src/statusbar.rs
use std::cell::{Cell, RefCell};
use std::rc::Rc;

use objc2::rc::Retained;
use objc2::runtime::NSObject;
use objc2::{define_class, msg_send, sel, DeclaredClass};
use objc2_app_kit::*;
use objc2_foundation::{MainThreadMarker, NSString};

use crate::gamepad::GamepadManager;
use crate::settings::Settings;
use crate::settings_window::SettingsWindow;

struct MenuTargetIvars {
    enabled: Rc<Cell<bool>>,
    enabled_item: RefCell<Option<Retained<NSMenuItem>>>,
    status_button: RefCell<Option<Retained<NSStatusBarButton>>>,
    settings: Rc<RefCell<Settings>>,
    settings_window: RefCell<SettingsWindow>,
}

define_class!(
    #[unsafe(super(NSObject))]
    #[name = "GpadMenuTarget"]
    #[ivars = MenuTargetIvars]
    struct MenuTarget;

    impl MenuTarget {
        #[unsafe(method(toggleEnabled:))]
        fn toggle_enabled(&self, _sender: &NSObject) {
            let iv = self.ivars();
            let was = iv.enabled.get();
            iv.enabled.set(!was);
            let now = !was;

            if let Some(item) = iv.enabled_item.borrow().as_ref() {
                item.setState(if now { 1 } else { 0 });
            }
            if let Some(button) = iv.status_button.borrow().as_ref() {
                let name = if now { "gamecontroller.fill" } else { "gamecontroller" };
                let image = NSImage::imageWithSystemSymbolName_accessibilityDescription(
                    &NSString::from_str(name),
                    Some(&NSString::from_str("gpad2mouse")),
                );
                if let Some(image) = image {
                    button.setImage(Some(&image));
                }
            }
            eprintln!("gpad2mouse: {}", if now { "enabled" } else { "disabled" });
        }

        #[unsafe(method(openSettings:))]
        fn open_settings(&self, _sender: &NSObject) {
            self.ivars()
                .settings_window
                .borrow_mut()
                .show(&self.ivars().settings);
        }
    }
);

pub struct StatusBar {
    _status_item: Retained<NSStatusItem>,
    _menu_target: Retained<MenuTarget>,
    enabled: Rc<Cell<bool>>,
}

impl StatusBar {
    pub fn new(
        mtm: MainThreadMarker,
        _gamepad: &Rc<GamepadManager>,
        settings: &Rc<RefCell<Settings>>,
    ) -> Self {
        let status_bar = NSStatusBar::systemStatusBar();
        let status_item = status_bar.statusItemWithLength(NSVariableStatusItemLength);
        let button = status_item.button(mtm);

        if let Some(ref button) = button {
            let image = NSImage::imageWithSystemSymbolName_accessibilityDescription(
                &NSString::from_str("gamecontroller.fill"),
                Some(&NSString::from_str("gpad2mouse")),
            );
            if let Some(image) = image {
                button.setImage(Some(&image));
            }
        }

        let enabled = Rc::new(Cell::new(true));

        let menu_target = mtm.alloc::<MenuTarget>().set_ivars(MenuTargetIvars {
            enabled: Rc::clone(&enabled),
            enabled_item: RefCell::new(None),
            status_button: RefCell::new(button),
            settings: Rc::clone(settings),
            settings_window: RefCell::new(SettingsWindow::new()),
        });
        let menu_target: Retained<MenuTarget> = unsafe { msg_send![super(menu_target), init] };

        let menu = NSMenu::new(mtm);

        let info_item = NSMenuItem::new(mtm);
        info_item.setTitle(&NSString::from_str("gpad2mouse"));
        info_item.setEnabled(false);
        menu.addItem(&info_item);
        menu.addItem(&NSMenuItem::separatorItem(mtm));

        let enabled_item = unsafe {
            NSMenuItem::initWithTitle_action_keyEquivalent(
                mtm.alloc(),
                &NSString::from_str("Enabled"),
                Some(sel!(toggleEnabled:)),
                &NSString::from_str(""),
            )
        };
        enabled_item.setState(1);
        unsafe { enabled_item.setTarget(Some(&menu_target)) };
        menu.addItem(&enabled_item);
        *menu_target.ivars().enabled_item.borrow_mut() = Some(enabled_item);

        let settings_item = unsafe {
            NSMenuItem::initWithTitle_action_keyEquivalent(
                mtm.alloc(),
                &NSString::from_str("Settings..."),
                Some(sel!(openSettings:)),
                &NSString::from_str(","),
            )
        };
        unsafe { settings_item.setTarget(Some(&menu_target)) };
        menu.addItem(&settings_item);

        menu.addItem(&NSMenuItem::separatorItem(mtm));

        let quit_item = unsafe {
            NSMenuItem::initWithTitle_action_keyEquivalent(
                mtm.alloc(),
                &NSString::from_str("Quit"),
                Some(sel!(terminate:)),
                &NSString::from_str("q"),
            )
        };
        menu.addItem(&quit_item);

        status_item.setMenu(Some(&menu));

        Self {
            _status_item: status_item,
            _menu_target: menu_target,
            enabled,
        }
    }

    pub fn is_enabled(&self) -> bool {
        self.enabled.get()
    }
}
